### PR TITLE
Renew the CFD props interface with the latest `interafaces`

### PR DESCRIPTION
**Pull request by arealclimber**
### DEVELOP
- [x]  PositionOpenModal
- [x]  OpenSubTab 
- [x]  OpenPositionItem
- [x]  UpdateFormModal
- [x]  PositionUpdatedModal
- [x]  PositionClosedModal
- [x]  HistorySubTab
- [x]  HistoryPositionItem
- [x]  HistoryPositionModal
- [x]  globalContext

### CHECK LIST

- [Naming convention](https://github.com/CAFECA-IO/WorkGuidelines/blob/main/technology/coding-convention/naming-convention.md) verification: checked
- Coding style verification: checked
- new Library: 0
- new Class / Component: 0
- new loop: 0
- new recursive: 0
- high risk: 0
- new sql: 0

### UML

- [Sequence Diagram 001 (interaction between user, frontend and backend)](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDSD00001.md)
- [Sequence Diagram 002 (background mechanism)](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDSD00002.md)
- [Component Diagram](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDCP00001.md)
- [Class Diagram](https://github.com/CAFECA-IO/Documents/blob/main/TBD/TBDCD00001.md)
- [Global context document](https://www.notion.so/Global-Context-public-d93ac7c941c94fb6b173d34a7bc4e959)
- [Table of interface](https://github.com/CAFECA-IO/EventReport/blob/main/meeting/20230106_tbd_mvp.md)

### CODE REVIEW

File: src/interfaces/tidebit_defi_background/apply_cfd_order.ts [hunk 1]
```diff
@@ -17,9 +17,10 @@ export type ApplyCFDOrderData =
   | IApplyUpdateCFDOrderData
   | IApplyCloseCFDOrderData;
 
+// TODO: IApplyCFDOrder['data'] is `any` type
 export interface IApplyCFDOrder {
   type: ICFDOrderType;
-  data: ApplyCFDOrderData;
+  data: ApplyCFDOrderData; // TODO: separate to 3 types
   signature: string;
 }
```
Comment (arealclimber) [inline src/interfaces/tidebit_defi_background/apply_cfd_order.ts:25]: - Separate the `IApplyCreateCFDOrder, update, close` by extending `IApplyCFDOrder`

File: src/components/position_open_modal/position_open_modal.tsx [hunk 2]
```diff
@@ -155,11 +217,30 @@ const PositionOpenModal = ({
   // const displayedTime = timestampToString(openCfdRequest?.createdTime ?? 0);
 
   const renewDataHandler = async () => {
+    const dataRenewal = getDummyDisplayApplyCreateCFDOrder(marketCtx.selectedTicker!.currency);
+    const creatCfdData = dataRenewal.data as IApplyCreateCFDOrderData;
```
Comment (arealclimber) [inline src/components/position_open_modal/position_open_modal.tsx:221]: - 跟 marketCtx 拿報價 (Open price)
- required margin = price x amount / leverage
- fee = price x amount x fee rate (marketCtx)

File: src/components/circular_progress_bar/circular_progress_bar.tsx [hunk 3]
```diff
@@ -28,7 +29,7 @@ const CircularProgressBar = ({
   className,
 }: ICircularProgressBarProps) => {
   const percentage = [(numerator / denominator) * 100];
-  const label = [`${numerator} H`];
+  // const label = [`${numerator} H`];
```
Comment (Luphia) [inline src/components/circular_progress_bar/circular_progress_bar.tsx:32]: - annotation format

File: src/components/history_position_modal/history_position_modal.tsx [hunk 4]
```diff
@@ -17,11 +17,13 @@ import {IClosedCFDDetails} from '../../interfaces/tidebit_defi_background/closed
...
 import {CFDClosedType} from '../../constants/cfd_closed_type';
 import {OrderState} from '../../constants/order_state';
+import {IDisplayAcceptedCFDOrder} from '../../interfaces/tidebit_defi_background/display_accepted_cfd_order';
 
 interface IHistoryPositionModal {
   modalVisible: boolean;
   modalClickHandler: () => void;
-  closedCfdDetails: IClosedCFDDetails;
+  closedCfdDetails: IDisplayAcceptedCFDOrder;
+  // closedCfdDetails: IClosedCFDDetails;
```
Comment (Luphia) [inline src/components/history_position_modal/history_position_modal.tsx:26]: - annotation format

File: src/components/history_sub_tab/history_sub_tab.tsx [hunk 5]
```diff
@@ -1,12 +1,59 @@
...
-    <div key={cfd.id}>{/* <HistoryPositionItem closedCfdDetails={cfd} /> */}</div>
-  ));
+
+  const toHistoryPositionItem = (cfds: IAcceptedCFDOrder[]) => {
+    const displayedHistoryPositionList = cfds.map(cfd => {
+      if (!cfd.closePrice) return;
+
+      const openValue = twoDecimal(cfd.openPrice * cfd.amount);
+      const closeValue = twoDecimal(cfd.closePrice * cfd.amount);
+      // TODO: (20230314 - Shirley)
```
Comment (Luphia) [inline src/components/history_sub_tab/history_sub_tab.tsx:21]: - ToDo subject

File: src/components/history_sub_tab/history_sub_tab.tsx [hunk 6]
```diff
@@ -1,12 +1,59 @@
...
+
+      const pnl: IPnL = {
+        type: value > 0 ? ProfitState.PROFIT : value < 0 ? ProfitState.LOSS : ProfitState.EQUAL,
+        value: cfd.closePrice ? Math.abs(cfd.closePrice - cfd.openPrice) * cfd.amount : 0,
+      };
+
+      return {...cfd, pnl, openValue, closeValue, positionLineGraph, suggestion};
+    });
+
+    // console.log('displayedHistoryPositionList', displayedHistoryPositionList);
```
Comment (Luphia) [inline src/components/history_sub_tab/history_sub_tab.tsx:41]: - annotation format

File: src/components/history_sub_tab/history_sub_tab.tsx [hunk 7]
```diff
@@ -1,12 +1,59 @@
...
+  const historyPositionList = toHistoryPositionItem(userCtx.closedCFDs).map(cfd =>
+    cfd ? (
+      <div key={cfd.id}>
+        <HistoryPositionItem closedCfdDetails={cfd} />
+      </div>
+    ) : null
+  );
+
+  // console.log('toHistoryPositionItem', toHistoryPositionItem());
+  // console.log('userCtx.closedCFDs', userCtx.closedCFDs);
```
Comment (Luphia) [inline src/components/history_sub_tab/history_sub_tab.tsx:55]: - annotation format

File: src/components/open_position_item/open_position_item.tsx [hunk 8]
```diff
@@ -7,67 +7,133 @@ import {
...
 interface IOpenPositionItemProps {
-  openCfdDetails: IOpenCFDDetails;
+  openCfdDetails: IDisplayAcceptedCFDOrder;
 }
 
 const OpenPositionItem = ({openCfdDetails, ...otherProps}: IOpenPositionItemProps) => {
   // if (longOrShort !== 'long' && longOrShort !== 'short') return <></>;
   // if (profitOrLoss !== 'profit' && profitOrLoss !== 'loss') return <></>; if (profitOrLoss !== 'profit' && profitOrLoss !== 'loss') return <></>;
   // if (ticker !== 'ETH' && ticker !== 'BTC') return <></>;
+  // console.log('openCfdDetails', openCfdDetails);
```
Comment (Luphia) [inline src/components/open_position_item/open_position_item.tsx:38]: - annotation format

File: src/components/open_position_item/open_position_item.tsx [hunk 9]
```diff
@@ -7,67 +7,133 @@ import {
...
+  // 傳param進來的地方要改正確的 dummy data
+  // TODO: toApplyUpdatedCFD function
+  // TODO: toApplyClosedCFD function
+
+  const toApplyUpdatedCFD = () => {
+    const dummyData = {};
+    return dummyData;
+  };
+
+  // from param or from user context
```
Comment (Luphia) [inline src/components/open_position_item/open_position_item.tsx:58]: - Info

File: src/components/open_position_item/open_position_item.tsx [hunk 10]
```diff
@@ -7,67 +7,133 @@ import {
...
+  const toApplyClosedCFD = (): IDisplayApplyCFDOrder => {
```
Comment (Luphia) [inline src/components/open_position_item/open_position_item.tsx:59]: - rename: toDisplayApplyCFDOrder

File: src/components/open_position_item/open_position_item.tsx [hunk 11]
```diff
@@ -7,67 +7,133 @@ import {
...
+        liquidationPrice: randomIntFromInterval(1000, 10000),
+        liquidationTime: Date.now() / 1000 + 86400, // openTimestamp + 86400
+      },
+      signature: '0x',
+      pnl: {
+        type: ProfitState.PROFIT,
+        value: 50,
+      },
+    };
+    return dummyData;
```
Comment (Luphia) [inline src/components/open_position_item/open_position_item.tsx:95]: - const result = {...} // not dummyData
- const quotation = {...} // ToDo: fetch from marketContext (20230315 Luphia)
- parse data from openCfdDetails

File: src/components/open_sub_tab/open_sub_tab.tsx [hunk 12]
```diff
@@ -1,15 +1,73 @@
...
+        value: Math.abs(pnlSoFar),
+      };
+
+      return {...cfd, openValue, positionLineGraph, suggestion, pnl};
+    });
+
+    return displayedOpenPositionList;
+  };
+
+  // console.log(toOpenPositionItem(openCFDs)); // Info: (20230314 - Shirley) `openCFDs` data from `display_accepted_cfd_order`
```
Comment (Luphia) [inline src/components/open_sub_tab/open_sub_tab.tsx:66]: - annotation format

File: src/components/update_form_modal/update_form_modal.tsx [hunk 13]
```diff
@@ -477,39 +496,56 @@ const UpdatedFormModal = ({
...
-                marketCtx.tickerLiveStatistics!.sellEstimatedFilledPrice * 1.25
-              )
-            : 99999,
-        // latestPnL: {
-        //   type: randomIntFromInterval(0, 100) <= 2 ? ProfitState.PROFIT : ProfitState.LOSS,
-        //   value: randomIntFromInterval(0, 1000),
-        // },
-      },
-    });
+    // FIXME: close modal
```
Comment (Luphia) [inline src/components/update_form_modal/update_form_modal.tsx:521]: - fix this

File: src/contexts/global_context.tsx [hunk 14]
```diff
@@ -121,17 +123,18 @@ export const dummyDataPositionClosedModal: IDataPositionClosedModal = {
     //   type: ProfitState.PROFIT,
     // },
   },
-  // openCfdDetails: getDummyDisplayAcceptedCFDOrder('ETH'),
-  // renewalDeadline: new Date('2023-02-24T17:00:00').getTime(),
-  // latestProps: getDummyApplyCloseCFDOrderData('ETH'),
+  // displayAcceptedCloseCFD: getDummyDisplayAcceptedCFDOrder('BTC'),
+  // displayApplyCloseCFD: getDummyDisplayApplyCreateCFDOrder('BTC'),
```
Comment (Luphia) [inline src/contexts/global_context.tsx:127]: - annotation format

File: src/contexts/global_context.tsx [hunk 15]
```diff
@@ -121,17 +123,18 @@ export const dummyDataPositionClosedModal: IDataPositionClosedModal = {
...
 };
 
 export const dummyDataPositionUpdatedModal: IDataPositionUpdatedModal = {
-  openCfdDetails: dummyOpenCFDDetails,
+  openCfdDetails: acceptedCFDOrders[0],
+  // openCfdDetails: getDummyDisplayApplyCreateCFDOrder('ETH'),
```
Comment (Luphia) [inline src/contexts/global_context.tsx:132]: - annotation format

File: src/contexts/global_context.tsx [hunk 16]
```diff
@@ -522,14 +522,16 @@ const initialColorMode: ColorModeUnion = 'dark';
...
   const userCtx = useContext(UserContext);
+  const marketCtx = useContext(MarketContext);
+
   const [colorMode, setColorMode] = useState<ColorModeUnion>(initialColorMode);
 
-  const [visiblePositionDetailsModal, setVisiblePositionDetailsModal] = useState(false);
+  const [visibleUpdateFormModal, setVisibleUpdateFormModal] = useState(false);
   // TODO: replace dummy data with standard example data
-  // const [dataPositionDetailsModal, setDataPositionDetailsModal] =
+  // const [dataUpdateFormModal, setDataUpdateFormModal] =
```
Comment (Luphia) [inline src/contexts/global_context.tsx:531]: - annotation format

File: src/contexts/global_context.tsx [hunk 17]
```diff
@@ -571,16 +573,20 @@ export const GlobalProvider = ({children}: IGlobalProvider) => {
...
 
   const [visiblePositionClosedModal, setVisiblePositionClosedModal] = useState(false);
   const [dataPositionClosedModal, setDataPositionClosedModal] = useState<IDataPositionClosedModal>(
     dummyDataPositionClosedModal
   );
 
   const [visiblePositionOpenModal, setVisiblePositionOpenModal] = useState(false);
+  // const [dataPositionOpenModal, setDataPositionOpenModal] = useState<IDisplayApplyCFDOrder>(
+  //   getDummyDisplayApplyCreateCFDOrder(marketCtx.selectedTickerRef.current!.currency)
+  // ); // TODO: Open position parameter
```
Comment (Luphia) [inline src/contexts/global_context.tsx:586]: - annotation format

File: src/contexts/global_context.tsx [hunk 18]
```diff
@@ -571,16 +573,20 @@ export const GlobalProvider = ({children}: IGlobalProvider) => {
...
   const [dataPositionOpenModal, setDataPositionOpenModal] = useState<IDataPositionOpenModal>(
     dummyDataPositionOpenModal
+    // getDummyDisplayApplyCreateCFDOrder('ETH')
```
Comment (Luphia) [inline src/contexts/global_context.tsx:589]: - annotation format

File: src/contexts/global_context.tsx [hunk 19]
```diff
@@ -1114,12 +1118,14 @@ export const GlobalProvider = ({children}: IGlobalProvider) => {
       <PositionOpenModal
         modalVisible={visiblePositionOpenModal}
         modalClickHandler={visiblePositionOpenModalHandler}
+        // displayApplyCreateCFD={dataPositionOpenModal.displayApplyCreateCFD}
```
Comment (Luphia) [inline src/contexts/global_context.tsx:1121]: - annotation format

File: src/contexts/global_context.tsx [hunk 20]
```diff
@@ -1114,12 +1118,14 @@ export const GlobalProvider = ({children}: IGlobalProvider) => {
...
         openCfdRequest={dataPositionOpenModal.openCfdRequest}
-        renewalDeadline={dataPositionOpenModal.renewalDeadline}
       />
       <PositionClosedModal
         modalVisible={visiblePositionClosedModal}
         modalClickHandler={visiblePositionClosedModalHandler}
+        // displayAcceptedCloseCFD={dataPositionClosedModal.displayAcceptedCloseCFD}
+        // displayApplyCloseCFD={dataPositionClosedModal.displayApplyCloseCFD}
```
Comment (Luphia) [inline src/contexts/global_context.tsx:1128]: - annotation format

**Review by Luphia — COMMENTED** [review]
- Fix these issues

File: src/components/position_closed_modal/position_closed_modal.tsx [hunk 21]
```diff
@@ -8,22 +8,50 @@ import {
...
 interface IPositionClosedModal {
   modalVisible: boolean;
   modalClickHandler: () => void;
-  openCfdDetails: IOpenCFDDetails;
+  // 送出時要 `displayApplyCloseCFD: IApplyCFDOrder;`
+  // TODO: 從 marketCtx 拿報價
+  // TODO: 自己算 PnL
+
+  openCfdDetails: IDisplayAcceptedCFDOrder;
   latestProps: IClosedCFDInfoProps;
```
Comment (arealclimber) [inline src/components/position_closed_modal/position_closed_modal.tsx:55]: 自己拿報價單

**Review by arealclimber — COMMENTED** [review]
fix these issues

**Review by Luphia — APPROVED** [review]
- no bug found

**Pull request merged**